identity-wallet/src/state/credentials/reducers/send_token_request.rs

```diff
@@ -79,9 +81,85 @@ pub async fn send_token_request(state: AppState, action: Action) -> Result<AppSt
 
         info!("credential issuer url: {:?}", credential_issuer_url);
 
+        // Get the credential issuer metadata.
+        let credential_issuer_metadata = wallet
+            .get_credential_issuer_metadata(credential_issuer_url.clone())
+            .await
+            .map_err(GetCredentialIssuerMetadataError)?;
+
+        // Check that the specified authorization servers are listed in the Credential Issuer Metadata's `authorization_servers` parameter.
+        if is_pre_authorized {
+            if let Some(grants) = &credential_offer.grants {
+                if let Some(pre_auth_code) = &grants.pre_authorized_code {
+                    if let Some(specified_auth_server) = &pre_auth_code.authorization_server {
+                        if !credential_issuer_metadata.authorization_servers.is_empty()
+                            && !credential_issuer_metadata
+                                .authorization_servers
+                                .contains(specified_auth_server)
+                        {
+                            return Err(AppError::Error(format!(
+                        "The specified authorization server {specified_auth_server} is not listed in the credential issuer metadata."
+                    )));
+                        }
+                    }
+                }
+            }
+        } else if let Some(grants) = &credential_offer.grants {
+            if let Some(auth_code) = &grants.authorization_code {
+                if let Some(specified_auth_server) = &auth_code.authorization_server {
+                    if !credential_issuer_metadata.authorization_servers.is_empty()
+                        && !credential_issuer_metadata
+                            .authorization_servers
+                            .contains(specified_auth_server)
+                    {
+                        return Err(AppError::Error(format!(
+                                "The specified authorization server {specified_auth_server} is not listed in the credential issuer metadata."
+                            )));
+                    }
+                }
+            }
+        }
```

Comment:
the `if` and `else` blocks are nearly identical so I think you should try to keep the duplicated code at a minimum :)